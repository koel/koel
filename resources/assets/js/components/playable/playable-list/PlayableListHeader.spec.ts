import { screen } from '@testing-library/vue'
import { ref } from 'vue'
import { describe, expect, it } from 'vitest'
import { createHarness } from '@/__tests__/TestHarness'
import {
  PlayableListConfigKey,
  PlayableListContextKey,
  PlayableListSortFieldKey,
  PlayableListSortOrderKey,
  SelectedPlayablesKey,
} from '@/symbols'
import PlayableListHeader from './PlayableListHeader.vue'

describe('playableListHeader.vue', () => {
  const h = createHarness()

  const renderComponent = async (
    config: Partial<PlayableListConfig> = {
      sortable: true,
      reorderable: true,
    },
    context: PlayableListContext = {
      type: 'Album',
    },
    selectedPlayables: Playable[] = [],
    sortField: PlayableListSortField = 'title',
    sortOrder: SortOrder = 'asc',
  ) => {
    const sortFieldRef = ref(sortField)
    const sortOrderRef = ref(sortOrder)

    h.visit('/songs')

    return h.render(PlayableListHeader, {
      props: {
        contentType: 'songs',
      },
      global: {
        stubs: {
          ActionMenu: h.stub(),
        },
        provide: {
          [<symbol>SelectedPlayablesKey]: [ref(selectedPlayables), (value: Playable[]) => (selectedPlayables = value)],
          [<symbol>PlayableListConfigKey]: [config],
          [<symbol>PlayableListContextKey]: [context],
          [<symbol>PlayableListSortFieldKey]: [sortFieldRef, (value: PlayableListSortField) => (sortFieldRef.value = value)],
          [<symbol>PlayableListSortOrderKey]: [sortOrderRef, (value: SortOrder) => (sortOrderRef.value = value)],
        },
      },
    })
  }

  it('renders', async () => {
    const { html } = await renderComponent()
    expect(html()).toMatchSnapshot()
  })

  it.each<[PlayableListSortField, string]>([
    ['track', 'header-track-number'],
    ['title', 'header-title'],
    ['album_name', 'header-album'],
    ['length', 'header-length'],
  ])('sorts by %s upon %s clicked', async (field, testId) => {
    const { emitted } = await renderComponent()

    await h.user.click(screen.getByTestId(testId))
    expect(emitted().sort[0]).toEqual([field, 'desc'])

    await h.user.click(screen.getByTestId(testId))
    expect(emitted().sort[1]).toEqual([field, 'asc'])
  })

  it('cannot be sorted if configured so', async () => {
    const { emitted } = await renderComponent({
      sortable: false,
      reorderable: true,
    })

    await h.user.click(screen.getByTestId('header-track-number'))
    expect(emitted().sort).toBeUndefined()
  })
})
