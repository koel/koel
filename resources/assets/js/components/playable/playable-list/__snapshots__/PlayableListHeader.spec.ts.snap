// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`playableListHeader.vue > renders 1`] = `
<div class="sortable song-list-header flex z-[2] bg-k-fg-3 pl-5"><span class="track-number" data-testid="header-track-number" role="button" title="Sort by track number"> # <!--v-if--><!--v-if--></span><span class="title-artist" data-testid="header-title" role="button" title="Sort by title"> Title <br data-testid="Icon" icon="[object Object]" class="text-k-highlight"><!--v-if--></span><span title="Sort by album" class="album" data-testid="header-album" role="button">Album<span class="ml-2"><!--v-if--><!--v-if--></span></span>
  <!--v-if-->
  <!--v-if-->
  <!--v-if--><span class="time" data-testid="header-length" role="button" title="Sort by duration"> Time <!--v-if--><!--v-if--></span><span class="extra"><article data-v-041d93ea=""><button data-v-041d93ea="" class="w-full focus:text-k-highlight" title="Sort"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></button><div data-v-041d93ea=""><menu data-v-041d93ea="" class="context-menu normal-case tracking-normal" style="display: none;"><li data-v-041d93ea="" class="active cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" checked="" disabled="" title="" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Title</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  <li data-v-041d93ea="" class="cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" checked="" title="Click to toggle the Album column" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Album</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  <li data-v-041d93ea="" class="cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" disabled="" title="" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Artist</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  <li data-v-041d93ea="" class="cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" checked="" title="Click to toggle the Track &amp; Disc column" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Track &amp; Disc</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  <li data-v-041d93ea="" class="cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" title="Click to toggle the Genre column" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Genre</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  <li data-v-041d93ea="" class="cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" title="Click to toggle the Year column" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Year</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  <li data-v-041d93ea="" class="cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" checked="" title="Click to toggle the Time column" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Time</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  <li data-v-041d93ea="" class="cursor-pointer flex justify-between !pl-3 hover:!bg-k-fg-10"><label data-v-041d93ea="" class="w-4 mr-2.5 flex items-center"><input data-v-041d93ea="" disabled="" title="" class="disabled:opacity-20 bg-white h-4 aspect-square rounded checked:border-white/75 checked:border-2 checked:bg-k-highlight" type="checkbox"></label><span data-v-041d93ea="" class="flex-1 text-left">Date Added</span><span data-v-041d93ea="" class="icon hidden ml-3"><br data-v-041d93ea="" data-testid="Icon" icon="[object Object]"></span></li>
  </menu>
</div>
</article></span></div>
`;
